libp2p/Common.cpp

```diff
@@ -180,25 +181,26 @@ Node::Node(NodeSpec const& _s, PeerType _p):
 
 NodeSpec::NodeSpec(string const& _user)
 {
-    m_address = _user;
-    if (m_address.substr(0, 8) == "enode://" && m_address.find('@') == 136)
+    // Format described here: https://github.com/ethereum/wiki/wiki/enode-url-format
+    // Example valid URLs:
+    //      enode://6332792c4a00e3e4ee0926ed89e0d27ef985424d97b6a45bf0f23e51f0dcb5e66b875777506458aea7af6f9e4ffb69f43f3778ee73c81ed9d34c51c4b16b0b0f@52.232.243.152:30305
+    //      enode://6332792c4a00e3e4ee0926ed89e0d27ef985424d97b6a45bf0f23e51f0dcb5e66b875777506458aea7af6f9e4ffb69f43f3778ee73c81ed9d34c51c4b16b0b0f@52.232.243.152:30305?discport=30303
+    //      enode://6332792c4a00e3e4ee0926ed89e0d27ef985424d97b6a45bf0f23e51f0dcb5e66b875777506458aea7af6f9e4ffb69f43f3778ee73c81ed9d34c51c4b16b0b0f@52.232.243.152
+    const char* peerPattern = "^(enode://)([\\dabcdef]{128})@(\\d{1,3}.\\d{1,3}.\\d{1,3}.\\d{1,3})((:\\d{2,5})(\\?discport=(\\d{2,5}))?)?$";
+    regex rx(peerPattern);
+    smatch match;
+
+    m_tcpPort = m_udpPort = c_defaultListenPort;
+    if (regex_match(_user, match, rx))
     {
-        m_id = p2p::NodeID(m_address.substr(8, 128));
-        m_address = m_address.substr(137);
-    }
-    size_t colon = m_address.find_first_of(":");
-    if (colon != string::npos)
-    {
-        string ports = m_address.substr(colon + 1);
-        m_address = m_address.substr(0, colon);
-        size_t p2 = ports.find_first_of(".");
-        if (p2 != string::npos)
+        m_id = p2p::NodeID(match.str(2));
+        m_address = match.str(3);
+        if (match[5].matched)
         {
-            m_udpPort = stoi(ports.substr(p2 + 1));
-            m_tcpPort = stoi(ports.substr(0, p2));
+            m_udpPort = m_tcpPort = stoi(match.str(5).substr(1));
```

Comment:
I guess with another set of parenthesis in regex you could get rid of `substr` call here

Reply:
I purposely used substr here since the regex was getting a bit unwieldy with all of the parentheses in the ports section...I'd like to leave this as-is unless you feel strongly about removing the substr.